### DISCUSSION TITLE
Does the App Router support PM2 cluster mode?

### DISCUSSION THREAD

**Discussion by developer-choi** [vercel/next.js#71946] [App Router]
https://github.com/developer-choi/test-pm2

I installed Next.js using `npx create-next-app@14.2.4`
and then added PM2 globally for testing. `npm i -g pm2`

However, I encountered the following error in cluster mode.

That every time I run That every time I run `pm2 reload ecosystem.config.js`, the service crashes, the service crashes, 
and when the service is down, running `pm2 reload ecosystem.config.js` brings it back to life.

![image](https://github.com/user-attachments/assets/c8dd1b00-09c8-4adf-b931-4c470155f1b8)

From my experience, out of 10 times I run `pm2 reload ecosystem.config.js`, the service crashes about 8 times. It’s not 100%, though.

As you can see, both pm2 and nextjs code were basic examples, but I had problems.

So I looked at the error log file. `pm2 logs`

![image](https://github.com/user-attachments/assets/d1b83802-6fab-4dbd-81b1-fa1d85eecbd3)

So I looked at pm2 and nextjs official documents, but I couldn't find the answer I wanted.

**_It doesn't have to be pm2. I want to do uninterrupted distribution._** What should I do?

**Comment by neo87cs** [comment]
Hi @developer-choi ,
We are using it in staging/production yet (next 14), i had to do some test to obtain the working version but looks like it is working with this ecosystem (in our case the script, with pnpm, has to be bypassed by pointing directly to next installed dist):
```
module.exports = {
  apps: [
    {
      args: 'start',
      cwd: process.env.PWD,
      env_development: {
        NODE_ENV: 'development',
      },
      env_production: {
        NODE_ENV: 'production',
        PORT: 3000,
      },
      env_staging: {
        NODE_ENV: 'production',
        PORT: 3000,
      },
      exec_mode: 'cluster',
      instances: 2,
      kill_timeout: 120000,
      listen_timeout: 10000,
      name: 'app',
      script: './node_modules/next/dist/bin/next',
    },
  ],
};
```
be aware that the default mode of pm2 ecosystem is `fork`, it means you will get 2 different servers running, if you want to achieve hot reload the correct exec mode is `cluster`

**Reply by quyphan97** [reply]
<img width="599" height="306" alt="image" src="https://github.com/user-attachments/assets/d1dffa90-9367-4bc2-b4b5-650ed3acd51f" />

Do you have any other solution? I updated it like that and it worked but the memory increased 56mb -> >500mb and continues to increase
<img width="971" height="205" alt="image" src="https://github.com/user-attachments/assets/1b323594-2fca-4ec1-8145-402b060622fc" />
